### PR TITLE
Added cli functionality, replaced apply function with a parametrical one

### PR DESCRIPTION
**Pull request by giacomocerquone**
Finally #4 I'm able to push this code. It took me a while mainly because I'm so fu***** busy!
I replaced the actual apply function and now both the cli.au3 and the function.au3 (not the gui.au3 as you said) call this function.
Now it needs absolutely some more tests and we have to provide also a shortcut to create links in the gui. If you can do it quickly it's ok, otherwise I can go there too.

PS. Just a quick question 'cause I couldn't find it online, what's this syntax with round parenthesis?
`
($array[0][0])[4)
`

### CODE REVIEW

**Comment by KurtisLiggett** [comment]
Thanks for the work you've done! I'll take a look at this soon.

> PS. Just a quick question 'cause I couldn't find it online, what's this syntax with round parenthesis?
($array[0][0])[4]

That is the syntax for an array of arrays.
In this example, $array has an array stored in [0][0], so we are accessing element [4] in the array $array[0][0].
*As you  can see, this ended up getting very messy and hard to follow, so I am reworking those parts of the program in my code cleanup...

**Comment by giacomocerquone** [comment]
I actually thought it was an array of arrays, but in javascript or C is 
`array[0][0][4]`
So it's the same

**Comment by KurtisLiggett** [comment]
I fixed a couple bugs.
I'll go ahead and add the'create link' functionality.
Just let me know if you already started working on it.

**Comment by giacomocerquone** [comment]
I started working on it but do it if you can!
You wrote the code then you're surely a lot faster than me!

**Comment by KurtisLiggett** [comment]
I added the link functionality. You can create a shortcut using the Tools menu or by right-clicking on a profile in the listview.
It should be fully functional, but error-checking needs to be improved.
Also considering some kind of status popup alerting the user that the script is actually running.

**Comment by giacomocerquone** [comment]
Everything seems good to me. Just two quick question and one note:

1. What is the variable $cmdline you defined as global? (it should be renamed and removed as global)
2. That StringInStr for the $sStdOut seems like an hack, coulnd't it be replaced with the correct error and not just searching for failed into the string?
3. could I replace that ugly `while 1` with this?

```Autoit
  While Not $pIdle
  	_asyncProcess()
  	Sleep(100)
  WEnd
  If StringInStr($sStdOut, "failed") Then
    MsgBox(1, "Error", "An error occurred."&@CRLF&@CRLF&$sStdOut)
  Else
    MsgBox(0,"Success", "Profile applied successfully.")
  EndIf
```

**Comment by KurtisLiggett** [comment]
> What is the variable $cmdline you defined as global? 

I put that in so we don't try to use any controls that haven't been created by the GUI yet.

> That StringInStr for the $sStdOut seems like an hack, coulnd't it be replaced with the correct error and not just searching for failed into the string?

Yes that is what I meant when I said error-checking needs to be improved.

> could I replace that ugly while 1 with this...

Yes please do that. But I am about to push some changes for the messages that you'll need to pull first.

> I think that the right place to put the "create link" button, would be on the toolbar with save, delete etc.

I agree. I moved it to the File menu.

> There is a bug saving a profile...

Good catch! I found a few more of them and fixed them.

**Comment by giacomocerquone** [comment]
Ok for the "while"!
I can't push until tomorrow and I can't see your changes, but I think I've been misunderstood. I'd like to put "Create Link" as an Icon (I call it toolbar because you called it so in the code)
I mean here: 
oie_c2xv2hzbebcv

**Comment by KurtisLiggett** [comment]
Oops, you're right, I didn't read what you said carefully enough.

I'm not sure about putting it in the toolbar since it isn't something people would be using every day.

**Comment by giacomocerquone** [comment]
It's true that people wouldn't use it everyday, but in the meantime it is a very (very very) important feature because in this way people wouldn't even open the app anymore. For this reason I think this should be highlighted.
But your call :)

**Comment by giacomocerquone** [comment]
Ok, one thing now! We should be consistent across notification! You're using this "Toast" library to show when the profile is applied, but when another instance of the programm is already running, you use the windows default notification system to notify the user.

Which one we wanna keep?

**Comment by KurtisLiggett** [comment]
Actually, when using the main GUI, the status bar shows messages for each action. If an error occurs, a preview of the error is displayed in the status bar. If you click on the message, the status bar expands to show the full message.

**Comment by giacomocerquone** [comment]
So what do you wanna do about the inconsistency of the notifications when applying a profile from a link and notifying when the program is already running?

Plus I think I have the solution: instead of placing an icon in the toolbar we could create a new link with a "drag 'n drop" from the profile to the directory where the user wants to place the link.
I can add the code right now if you think it's a proper solution

**Comment by KurtisLiggett** [comment]
I suppose to make things consistent, we should make the "already running" message a toast popup.  
 
--
> instead of placing an icon in the toolbar we could create a new link with a "drag 'n drop" from the profile to the directory where the user wants to place the link.

I like this!

**Comment by KurtisLiggett** [comment]
@giacomocerquone Just verifying.. you volunteered to add the drag 'n drop feature, correct?

**Comment by giacomocerquone** [comment]
Yeah of course! But as of now I don't know how much time I can spend working on this feature. Would you like to have it on the app shortly or it doesn't matter?

**Comment by KurtisLiggett** [comment]
No rush, I'm going to merge what we have now. Then you can add the drag and drop as a new feature when you get time.
